compiler/rustc_parse/src/parser/item.rs

```diff
@@ -2767,9 +2767,16 @@ impl<'a> Parser<'a> {
         // might be typo'd trait impl, handled elsewhere
         && !self.token.is_keyword(kw::For)
         {
-            // recover from missing argument list, e.g. `fn main -> () {}`
-            self.dcx()
-                .emit_err(errors::MissingFnParams { span: self.prev_token.span.shrink_to_hi() });
+            if self.token == token::PathSep && self.look_ahead(1, |t| *t == TokenKind::Lt) {
```

Comment:
I don't think this is the right place to insert this recovery. Specifically, we should be trying to parse the erroneous `::` before we've parsed the generic args list, so we can recover `fn foo::<T>(_: i32)` as `fn foo<T>(_: i32)`. Right now this recovers it as `fn foo()`.

Reply:
> I don't think this is the right place to insert this recovery. Specifically, we should be trying to parse the erroneous `::` before we've parsed the generic args list, so we can recover `fn foo::<T>(_: i32)` as `fn foo<T>(_: i32)`. Right now this recovers it as `fn foo()`.

Thank you.  I check method `parse_generics` before `parse_fn_params`, if the first token is not `token::Lt`, it will do nothing. So when beginning parse fn params we can find first token is `token::PathSep` and report the diagnostic info.

I tried this example `fn foo::<T>(_: i32)` and get following expected reports:

```rust
error: invalid path separator in function definition
 --> ./src/main.rs:7:7
  |
7 | fn foo::<T>(_: i32){};
  |       ^^
  |
help: remove invalid path separator
  |
7 - fn foo::<T>(_: i32){};
7 + fn foo<T>(_: i32){};
```

Reply:
> I don't think this is the right place to insert this recovery. Specifically, we should be trying to parse the erroneous `::` before we've parsed the generic args list, so we can recover `fn foo::<T>(_: i32)` as `fn foo<T>(_: i32)`. Right now this recovers it as `fn foo()`.

Thank you. I thought again and I agreed the correct place to find this problem and report the error should be in `parse_generics`. Please review again for me.